### PR TITLE
Feat/tx routing

### LINKED ISSUE #551: Transaction Routing UI
Create a UI to display transaction routing

First step
Screen Shot 2021-07-12 at 10 01 10 PM

Tooltip
Screen Shot 2021-07-12 at 10 01 39 PM

### PR DESCRIPTION
**Pull request by lijac**
Close #551

### CODE REVIEW

**Comment by octocat512** [comment]
- [ ] integrate testnet contract.
- [ ] add sICX/ICX pair to the calculation of route.
- [ ] refactor swap panel using new data structure.
- [ ]  refactor adding/removing liquidity
- [ ] refactor sendpanel component using new data structure

**Comment by octocat512** [comment]
https://github.com/balancednetwork/balanced-network-interface/issues/551

**Comment by gedanziger** [comment]
yeiodo address is here: cx4c456f4a02d2576fe712ea10b311a5fe8d06d205

Note that ICX swaps are supported now. This contract can also be used to stake ICX. Just specify a swap of ICX to SICX, and you'll get sICX back. If you specify a longer path, you can swap through a few different things.

You can also send back `null` as one of the addresses in the path. If the token type directly preceeding it was sICX, then you can swap from sICX to ICX this way.

It will fail if you do more than 10 swaps in your path.

Here's 2 modes of calling it:

1: sending in ICX:

```
path = [contracts['sicx']['SCORE'], contracts['bnUSD']['SCORE']]

params = {'_path': path}

print(json.dumps(params))
transaction = CallTransactionBuilder()\
    .from_(wallet.get_address())\
    .to(contracts['router']['SCORE'])\
    .value(10 * ICX)\
    .step_limit(10000000)\
    .nid(NID)\
    .nonce(100)\
    .method("route")\
    .params(params)\
    .build()
signed_transaction = SignedTransaction(transaction, wallet)
tx_hash = icon_service.send_transaction(signed_transaction)
tx_hash
```

2: sending in an IRC2

```
path = [contracts['bnUSD']['SCORE'], contracts['sicx']['SCORE'], contracts['bnUSD']['SCORE'], contracts['sicx']['SCORE'], None]

path_json = json.dumps(path)


to_token = contracts['bnUSD']['SCORE']
params_data = "{\"method\": \"_swap\", \"params\": {\"path\":" + path_json + "}}"
data = params_data.encode("utf-8")
params = {'_to': contracts['router']['SCORE'], '_value': ICX // 100, '_data': data}

print(params)
transaction = CallTransactionBuilder()\
    .from_(wallet.get_address())\
    .to(contracts['sicx']['SCORE'])\
    .value(0)\
    .step_limit(10000000)\
    .nid(NID)\
    .nonce(100)\
    .method("transfer")\
    .params(params)\
    .build()
signed_transaction = SignedTransaction(transaction, wallet)
tx_hash = icon_service.send_transaction(signed_transaction)
tx_hash

res = get_tx_result(tx_hash)
print(f'Status: {res["status"]}')
if len(res["eventLogs"]) > 0:
    for item in res["eventLogs"]:
        print(f'{item} \n')
if res['status'] == 0:
    print(f'Failure: {res["failure"]}')
```

**Comment by BennyOptions** [comment]
two methods:

tokenFallback - token transfer to the router
- pass between any token, but not ICX - only sICX can convert to ICX

ICXTransfer - router executes trade
- first tx in the path must convert to sICX

Priorities
- Check if an exact pool exists
- If exists, just trade from first token to second token
- If the path does not exist, check if bnUSD path exists
- If the bnUSD path does not exist, check for sICX path

@gedanziger please correct this if there is information I missed or is incorrect

**Comment by octocat512** [comment]
How to calculate the price? We calculate the execution price of each trade. why we need this price? I'd like to remove the price.
@BennyOptions @parrot9design 
image

I'd like to move the execution price data on the swap panel.

image

**Comment by octocat512** [comment]
I also like to remove the sICX/ICX chart.

File: src/app/components/trade/SwapPanel.tsx [hunk 1]
```diff
@@ -144,67 +144,45 @@ export default function SwapPanel() {
...
-          .finally(() => {
-            window.removeEventListener('beforeunload', showMessageOnBeforeUnload);
-            changeShouldLedgerSign(false);
-          });
-      }
+    if (executionTrade.inputAmount.currency.symbol === 'ICX') {
+      bnJs
+        .inject({ account })
+        .Router.swapICX(
+          BalancedJs.utils.toLoop(new BigNumber(executionTrade.inputAmount.toExact())),
```
Comment (octocat512) [inline src/app/components/trade/SwapPanel.tsx:151]: `toLoop` can accept the string value. so please use `executionTrade.inputAmount.toFixed()`.

**Review by octocat512 — CHANGES_REQUESTED** [review]

**Comment by parrot9design** [comment]
- [ ] Re-add the ICX/sICX chart, so people can always see that the sICX value is increasing over time compared to ICX
- [ ] Re-add the last execution price to all available price charts
- [ ] Fix the price charts for OMM/sICX, OMM/IUSDC, and OMM/USDS

---
It seems that the amount displayed in the banner message is always the same as the value entered in the `For` field, even if the final amount received from the swap is different.

For example: 

Screen Shot 2021-09-29 at 6 28 36 PM

This swap was 200 ICX for 235.21 IUSDC. Before I initiated this transaction, my IUSDC balance was 0. It says I received 235.21 IUSDC, but the UI and [transaction](https://bicon.tracker.solidwallet.io/transaction/0x2a0c1b261b7b26a8a85e2a9762ad33a1b0a228167776bbef7dd74855f300b28b) indicate that it was actually 219.62168 IUSDC. I'm unable to check my activity history on the testnet, so I can't see how it's being reported there.

**Comment by BennyOptions** [comment]
blocked by ICON2 migration

**Comment by lijac** [comment]
- [x]  Re-add the ICX/sICX chart, so people can always see that the sICX value is increasing over time compared to ICX
- [x]  Re-add the last execution price to all available price charts
- [x]  Fix the price charts for OMM/sICX, OMM/IUSDC, and OMM/USDS
         Works on the production environment, caused by not having historical data in the test environment.

The bug - not reporting the correct amount - is fixed.

**Comment by octocat512** [comment]
the network id = 83.
you can check the pool ids on the master branch.

**Comment by Chunn** [comment]
> Please refactor the `TransactionPanel` according to the new constant data. I am gonna remove balanced non-related tokens in addresses.ts.(IUSDT, IUSDC, METX, OMM, CFT) And will remove `CURRENCY_INFO` and `CURRENCY` in the `src/constant/currency.ts`. @rise-techiast 11
> 
> * [ ]  remove hard-coded `POOL_IDS` . use the SUPPORTED_PAIRS_INFO constant in `/src/constants/pairs.ts`.
> * [ ]  don't use addresses.ts. use the SUPPORTED_TOKENS_LIST constant in `/src/constants/tokens.ts`
>   https://github.com/balancednetwork/balanced-network-interface/blob/c29e307c4da9bf50f1b24991c7da99c26acd75f9/src/app/components/home/TransactionPanel.tsx#L93
> * [ ]  don't use `CURRENCY` constant


@octocat512 (@rise-techiast ) I've updated based on your comment. However, I cannot push my branch to 'https://github.com/lijac/balanced-network-interface.git'. Could you grant me permission to access this repo? Thank you!

Screenshot 2021-11-01 233028

**Comment by octocat512** [comment]
Please use the GitHub CLI to work on the branch.
https://cli.github.com/
To pull the branch, use this command.
`gh pr checkout 657`

**Comment by Chunn** [comment]
@octocat512 (@rise-techiast ) I've refactored the transaction panel. Please check: https://github.com/balancednetwork/balanced-network-interface/pull/657/commits/de759ab2d0c71bb41565669a622ddc01cb7b66d8

**Comment by octocat512** [comment]
Thanks for your refactoring. @Chunn 
But the UI logic is not fit for scalability. If we add more tokens like USDC or IUSDT(6 decimals tokens), then we have to fix the logic. The refactoring goal is to not touch the TransactionPanel UI logic while adding new tokens.
for example.
image

So I'd like you to refactor the code again. Use the Token data structure and `parseUnits`/`formatUnits` for formatting.
- remove token-specific functions(isIUSDC, isIUSDT, convertIUSDC) and make a generalized converting function(convertValue). `convertValue(value: string, address: string)`. you can get token information by using its address. Then you can get decimal info of its.

image
image
 
image
image
Please use the generalized `convertValue` function that you will create in the above sections.

**Comment by octocat512** [comment]
The tx routing will be live next week. So I'd like you to finish the refactoring this week.

**Comment by Chunn** [comment]
@octocat512 (@rise-techiast ) I've refactored transaction panel again. Please check https://github.com/balancednetwork/balanced-network-interface/pull/657/commits/f288d53ee827bc58bcad098efbc783751a9f513f. Thank you!

File: src/app/components/home/TransactionPanel.tsx [hunk 2]
```diff
@@ -112,8 +107,14 @@ const AmountItem = ({ value, symbol, positive }: { value?: string; symbol?: stri
   </Typography>
 );
 
-const convertValue = (value: string) =>
-  BalancedJs.utils.toIcx(value).isGreaterThan(0.004) ? formatBigNumber(BalancedJs.utils.toIcx(value), 'currency') : '';
+const convertValue = (value: string, currencyKey?: string) => {
+  const currency = BalancedJs.utils.toIcx(value, currencyKey);
+  const exceptionList = ['IUSDT', 'IUSDC'];
```
Comment (octocat512) [inline src/app/components/home/TransactionPanel.tsx:112]: Please don't use the `toICX` and `toLoop` functions. will be deprecated.

File: src/app/components/home/TransactionPanel.tsx [hunk 3]
```diff
@@ -348,30 +307,11 @@ const getAmountWithSign = (tx: Transaction) => {
...
-        <>
-          <AmountItem value={amount1} symbol={symbol1} positive={true} />
-          <AmountItem value={amount2} symbol={symbol2} positive={true} />
-          <AmountItem value={amount3} symbol={symbol3} positive={true} />
-          <AmountItem value={amount4} symbol={symbol4} positive={true} />
-          <AmountItem value={amount5} symbol={symbol5} positive={true} />
-          <AmountItem value={amount6} symbol={symbol6} positive={true} />
-        </>
-      );
+      const amountItemList = getValuesAndSymbols(tx);
```
Comment (octocat512) [inline src/app/components/home/TransactionPanel.tsx:310]: There are two `Claimed` cases. Please merge the logic in one case.

**Review by octocat512 — CHANGES_REQUESTED** [review]

File: src/app/components/home/TransactionPanel.tsx (hunk 3 above)
Comment (Chunn) [reply src/app/components/home/TransactionPanel.tsx:310]: @octocat512 one for getContent() one for getValuesAndSymbols(). Please check again. Thank you!

Comment (octocat512) [reply src/app/components/home/TransactionPanel.tsx:310]: haha. my bad. @Chunn

File: src/constants/tokens.ts [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+  SupportedChainId.MAINNET, //
+  'cx369a5f4ce4f4648dfc96ba0c8229be0693b4eca2',
+  18,
+  'METX',
+  'Metanyx',
+);
+export const IUSDT = new Token(
+  SupportedChainId.MAINNET, //
+  'cx3a36ea1f6b9aa3d2dd9cb68e8987bcc3aabaaa88',
+  18,
```
Comment (Chunn) [inline src/constants/tokens.ts:67]: @octocat512 It seems IUSDT decimals should be 6

Comment (octocat512) [reply src/constants/tokens.ts:67]: Good Catch! Thanks

**Comment by Chunn** [comment]
@octocat512 I've updated. Please check https://github.com/balancednetwork/balanced-network-interface/pull/657/commits/697d6b4b7e73209e692f1fb1ef0f40784c34c754. Thanks!

**Comment by parrot9design** [comment]
Only the ICX logo is being displayed right now, is that expected?
Screen Shot 2021-11-09 at 10 04 30 AM

**Comment by octocat512** [comment]
> Only the ICX logo is being displayed right now, is that expected? Screen Shot 2021-11-09 at 10 04 30 AM

yeah, I only added the token logos for mainnet. Let me add logos for testnet.

**Comment by parrot9design** [comment]
Can we please re-add the spaces for liquidity pools (before and after the `/`)? For some reason they've been removed in this PR, which makes it harder to read.

**This PR** | **App**
----------- | ---------
Screen Shot 2021-11-16 at 5 08 55 PM | Screen Shot 2021-11-16 at 5 10 12 PM

**Comment by BennyOptions** [comment]
Current issues 

- [ ] Supply liquidity issue (slider does not reflect textbox inputs
- [ ] Price impact is incorrect for all trades

**Comment by lijac** [comment]
balanced-network-interface/src/app/components/trade/SupplyLiquidityModal.tsx L81 contains an error.

We have removed token contracts from bnJs, so individual token contracts (except sICX, ICX, BALN, bnUSD) aren't available.

**Comment by BennyOptions** [comment]
Ok, it seems that price impact calculation is fixed

**Pull request merged**